File: src/main/java/com/airsaid/localization/ui/SupportLanguagesDialog.java
/*
 * Copyright 2021 Airsaid. https://github.com/airsaid
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.airsaid.localization.ui;

import com.airsaid.localization.translate.AbstractTranslator;
import com.airsaid.localization.translate.lang.Lang;
import com.intellij.openapi.ui.DialogWrapper;
import com.intellij.ui.components.JBLabel;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.awt.*;
import java.util.Comparator;
import java.util.List;

/**
 * @author airsaid
 */
public class SupportLanguagesDialog extends DialogWrapper {

  private final AbstractTranslator mTranslator;

  public SupportLanguagesDialog(AbstractTranslator translator) {
    super(true);
    setTitle(translator.getName() + " Translator Supported Languages");
    mTranslator = translator;
    init();
  }

  @Override
  protected @Nullable JComponent createCenterPanel() {
    List<Lang> supportedLanguages = mTranslator.getSupportedLanguages();
    supportedLanguages.sort(new EnglishNameComparator());
    JPanel contentPanel = new JPanel(new GridLayout(supportedLanguages.size() / 4, 4, 10, 20));
    for (Lang supportedLanguage : supportedLanguages) {
      contentPanel.add(new JBLabel(supportedLanguage.getEnglishName()));
    }
    return contentPanel;
  }

  @Override
  protected @Nullable String getDimensionServiceKey() {
    String key = mTranslator.getKey();
    return "#com.airsaid.localization.ui.SupportLanguagesDialog#".concat(key);
  }

  @Override
  protected Action @NotNull [] createActions() {
    return new Action[]{};
  }

  static class EnglishNameComparator implements Comparator<Lang> {
    @Override
    public int compare(Lang o1, Lang o2) {
      return o1.getEnglishName().compareTo(o2.getEnglishName());
    }
  }
}

File: src/main/kotlin/com/airsaid/localization/ui/SupportLanguagesDialog.kt
/*
 * Copyright 2021 Airsaid. https://github.com/airsaid
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.airsaid.localization.ui

import androidx.compose.foundation.VerticalScrollbar
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.rememberScrollbarAdapter
import androidx.compose.ui.Alignment
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.awt.ComposePanel
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.airsaid.localization.translate.AbstractTranslator
import com.airsaid.localization.translate.lang.Lang
import com.airsaid.localization.translate.lang.flagEmoji
import com.intellij.openapi.ui.DialogWrapper
import javax.swing.Action
import javax.swing.JComponent

class SupportLanguagesDialog(private val translator: AbstractTranslator) : DialogWrapper(true) {

    private val supportedLanguages: List<Lang> = translator.supportedLanguages.sortedBy { it.englishName }

    init {
        title = "${translator.name} Translator Supported Languages"
        init()
    }

    override fun createCenterPanel(): JComponent {
        val panel = ComposePanel()
        panel.preferredSize = java.awt.Dimension(460, 420)
        panel.setContent {
            IdeTheme {
                Surface(
                    modifier = Modifier.fillMaxWidth(),
                    color = MaterialTheme.colorScheme.background,
                ) {
                    SupportLanguagesContent(languages = supportedLanguages)
                }
            }
        }
        return panel
    }

    override fun getDimensionServiceKey(): String? {
        val key = translator.key
        return "#com.airsaid.localization.ui.SupportLanguagesDialog#$key"
    }

    override fun createActions(): Array<Action> = emptyArray()
}

@Composable
private fun SupportLanguagesContent(languages: List<Lang>) {
    val listState = rememberLazyListState()

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .background(MaterialTheme.colorScheme.background)
            .padding(horizontal = 20.dp, vertical = 24.dp)
    ) {
        Text(
            text = "Supported languages (${languages.size})",
            style = MaterialTheme.typography.titleSmall,
            color = MaterialTheme.colorScheme.onBackground
        )
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .padding(top = 12.dp)
        ) {
            LazyColumn(
                modifier = Modifier.fillMaxSize(),
                state = listState,
                verticalArrangement = Arrangement.spacedBy(12.dp)
            ) {
                items(languages, key = { it.id }) { language ->
                    Column(modifier = Modifier.fillMaxWidth()) {
                        Row(
                            verticalAlignment = Alignment.CenterVertically,
                            horizontalArrangement = Arrangement.spacedBy(8.dp)
                        ) {
                            val flag = language.flagEmoji
                            if (flag != null) {
                                Text(
                                    text = flag,
                                    style = MaterialTheme.typography.headlineSmall.copy(fontSize = 24.sp)
                                )
                            }
                            Text(text = language.englishName, style = MaterialTheme.typography.bodyMedium)
                        }
                        Text(
                            text = language.code.uppercase(),
                            style = MaterialTheme.typography.labelSmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant
                        )
                    }
                }
            }
            VerticalScrollbar(
                modifier = Modifier.align(Alignment.CenterEnd),
                adapter = rememberScrollbarAdapter(listState)
            )
        }
    }
}
